### PR TITLE
[SPARK-12817] Add BlockManager.getOrElseUpdate and remove CacheManager

### PR DESCRIPTION
**Pull request by JoshRosen**
CacheManager directly calls MemoryStore.unrollSafely() and has its own logic for handling graceful fallback to disk when cached data does not fit in memory. However, this logic also exists inside of the MemoryStore itself, so this appears to be unnecessary duplication.

Thanks to the addition of block-level read/write locks in #10705, we can refactor the code to remove the CacheManager and replace it with an atomic `BlockManager.getOrElseUpdate()` method.

This pull request replaces / subsumes #10748.

/cc @andrewor14 and @nongli for review. Note that this changes the locking semantics of a couple of internal BlockManager methods (`doPut()` and `lockNewBlockForWriting`), so please pay attention to the Scaladoc changes and new test cases for those methods.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/CacheManagerSuite.scala [hunk 1]
```diff
@@ -1,97 +0,0 @@
...
-        firstParent[Int].iterator(split, context)
-    }.cache()
-    rdd3 = new RDD[Int](sc, List(new OneToOneDependency(rdd2))) {
-      override def getPartitions: Array[Partition] = firstParent[Int].partitions
-      override def compute(split: Partition, context: TaskContext): Iterator[Int] =
-        firstParent[Int].iterator(split, context)
-    }.cache()
-  }
-
-  test("get uncached rdd") {
```
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/CacheManagerSuite.scala]: Open question: which test cases do we need to preserve in the new code? Should I test this more end-to-end to make sure that caching still works (i.e. doesn't recompute + stores blocks as side effect), or is that handled at a higher level of the stack in existing tests?

File: core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala [hunk 2]
```diff
@@ -307,29 +307,48 @@ private[storage] class BlockInfoManager extends Logging {
...
+        // If there is no blockInfo for the given block, then the block does not exist and the
+        // caller won the race to write this block, so create a new blockInfo and lock the block
+        // for writing.
+        infos(blockId) = newBlockInfo
+        newBlockInfo.writerTask = currentTaskAttemptId
+        writeLocksByTask.addBinding(currentTaskAttemptId, blockId)
+        logTrace(s"Task $currentTaskAttemptId successfully locked new block $blockId")
+        return true
+      } else {
+        // Otherwise, the block already exists. If the block is locked for writing, however, it's
```
Comment (nongli) [inline core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala]: The second sentence in this comment is hard to parse.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 3]
```diff
@@ -648,6 +647,32 @@ private[spark] class BlockManager(
...
   /**
+   * Retrieve the given block if it exists, otherwise call the provided `makeIterator` method
+   * to compute the block, persist it, and return its values.
+   */
+  def getOrElseUpdate(
+      blockId: BlockId,
+      level: StorageLevel,
+      makeIterator: () => Iterator[Any]): BlockResult = {
+    // Initially we hold no locks on this block.
+    doPut(blockId, IteratorValues(makeIterator), level, downgradeToReadLock = true)
```
Comment (nongli) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: does anyone currently check the return value of doPut()? What if just returned Option[BlockResult] and then you can skip some of the logic below.

**Comment by nongli** [comment]
much simpler!

**Comment by JoshRosen** [comment]
There's one corner-case which this patch doesn't handle correctly: if we are trying to cache a block, start unrolling the iterator, and then conclude that the iterator won't fit in memory **and** we can't drop it down to disk, then we need to have a way to keep the values that we've already computed. TL;DR: the memoryStore itself now needs a `maybePutAndReturnIterator` method in order to preserve the old behavior.

**Comment by JoshRosen** [comment]
Alright, took another shot at addressing the correctness issue. Will address other comments after tests pass.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 4]
```diff
@@ -852,18 +878,20 @@ private[spark] class BlockManager(
             Await.ready(replicationFuture, Duration.Inf)
           }
         case _ =>
-          val remoteStartTime = System.currentTimeMillis
-          // Serialize the block if not already done
-          if (bytesAfterPut == null) {
-            if (valuesAfterPut == null) {
-              throw new SparkException(
-                "Underlying put returned neither an Iterator nor bytes! This shouldn't happen.")
+          if (blockWasSuccessfullyStored) {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: /cc @tdas, the goal of this change is to avoid attempting to replicate deserialized, memory-only blocks if their initial cache / persist fails due to a lack of memory.

For the purposes of this patch, we need to do this to prevent the iterator from being consumed so that it can be passed back to the caller. More generally, though, I think that we should have this change to avoid OOMs by trying to serialize an entire partition which was too large to be stored.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 5]
```diff
@@ -315,6 +315,34 @@ abstract class RDD[T: ClassTag](
   }
 
   /**
+   * Gets or computes an RDD partition. Used by RDD.iterator() when an RDD is cached.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Can you add that this is only called on the executors? (otherwise people might wonder why we don't just use `sc.env` instead of `SparkEnv.get`)

File: core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala [hunk 6]
```diff
@@ -307,29 +307,48 @@ private[storage] class BlockInfoManager extends Logging {
...
-      infos(blockId) = newBlockInfo
-      newBlockInfo.writerTask = currentTaskAttemptId
-      writeLocksByTask.addBinding(currentTaskAttemptId, blockId)
-      logTrace(s"Task $currentTaskAttemptId successfully locked new block $blockId")
-      true
-    } else {
-      logTrace(s"Task $currentTaskAttemptId did not create and lock block $blockId " +
-        s"because that block already exists")
-      false
+    while(true) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala]: space after while

File: core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala [hunk 7]
```diff
@@ -307,29 +307,48 @@ private[storage] class BlockInfoManager extends Logging {
...
+        writeLocksByTask.addBinding(currentTaskAttemptId, blockId)
+        logTrace(s"Task $currentTaskAttemptId successfully locked new block $blockId")
+        return true
+      } else {
+        // Otherwise, the block already exists. If the block is locked for writing, however, it's
+        // possible that the writer might delete the block or might be in the process of writing the
+        // block (which could fail); in this case, we choose to block until that write lock is
+        // released so we know whether the block is available for reading before returning. This
+        // semantic is useful in the implementation of the BlockManager's getOrElseUpdate() method.
+        if (lockForReading(blockId).isDefined) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala]: This works, but I think there's a cleaner way to do the same thing without the while loop:
```
/**
 * Attempt to acquire the appropriate lock for writing a new block.
 *
 * This enforces the first-writer-wins semantics. If we are the first to write the block,
 * then just go ahead and acquire the write lock. Otherwise, if another thread is already
 * writing the block, then we wait for the write to finish before acquiring the read lock.
 *
 * [... explain return value here ...]
 */
def lockNewBlockForWriting(blockId: BlockId, blockInfo: BlockInfo): Boolean = synchronized {
  lockForReading(blockId) match {
    case Some(binfo) =>
      // Block already exists. This could happen if another thread races with us to compute
      // the same block. In this case, just keep the read lock and return.
      false
    case None =>
      // Block does not yet exist or is removed, so we are free to acquire the write lock
      infos(blockId) = blockInfo
      lockForWriting(blockId)
      true
  }
}
```
To me this is much easier to explain than what you currently have.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 8]
```diff
@@ -707,26 +719,33 @@ private[spark] class BlockManager(
...
-   * The effective storage level refers to the level according to which the block will actually be
-   * handled. This allows the caller to specify an alternate behavior of doPut while preserving
-   * the original level specified by the user.
-   *
-   * @return true if the block was stored or false if the block was already stored or an
-   *         error occurred.
+   * @param effectiveStorageLevel the level according to which the block will actually be handled.
+   *                              This allows the caller to specify an alternate behavior of doPut
+   *                              while preserving the original level specified by the user.
+   * @param downgradeToReadLock if true, this method will downgrade its write lock on the block
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I think a better name is `returnWithReadLock`. If the block already exists, we're not really "downgrading" it, since we never owned the write lock in the first place. Then the comment to explain it can be clearer:
```
If true, this method will hold the read lock when it returns even if the block already exists.
If false, this method will hold no locks when it returns.
```

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 9]
```diff
@@ -707,26 +719,33 @@ private[spark] class BlockManager(
       tellMaster: Boolean = true,
       effectiveStorageLevel: Option[StorageLevel] = None): Boolean = {
     require(bytes != null, "Bytes is null")
-    doPut(blockId, ByteBufferValues(bytes), level, tellMaster, effectiveStorageLevel)
+    doPut(blockId, ByteBufferValues(bytes), level, tellMaster, effectiveStorageLevel).isEmpty
   }
 
   /**
    * Put the given block according to the given level in one of the block stores, replicating
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: This needs to be updated to reflect that we don't actually put the block if someone else has already put it.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 10]
```diff
@@ -707,26 +719,33 @@ private[spark] class BlockManager(
...
+   *                              while preserving the original level specified by the user.
+   * @param downgradeToReadLock if true, this method will downgrade its write lock on the block
+   *                            to a read lock before returning; this will happen even if the block
+   *                            already exists, so when this is true the caller will always hold
+   *                            a read lock on the block after this method returns without throwing
+   *                            an exception. If false (default), this method will release the write
+   *                            lock before returning.
+   * @return `Some(PutResult)` if the block did not exist and could not be successfully cached,
+   *         or None if the block already existed or was successfully stored (fully consuming
+   *         the input data / input iterator).
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I read this many times and still found this confusing. Normally I would expect to get `Some`thing when the operation succeeded, and `None` otherwise, but this is the other way round. I think it would be simpler if we add a field in `PutResult` that says `success: Boolean` or something and just always return a `PutResult`. Then you don't need to get the block again after putting in `getOrElseUpdate`, which I also found kind of unintuitive.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 11]
```diff
@@ -648,8 +647,38 @@ private[spark] class BlockManager(
...
+   *
+   * @return either a BlockResult if the block was successfully cached, or an iterator if the block
+   *         could not be cached.
+   */
+  def getOrElseUpdate(
+      blockId: BlockId,
+      level: StorageLevel,
+      makeIterator: () => Iterator[Any]): Either[BlockResult, Iterator[Any]] = {
+    // Initially we hold no locks on this block.
+    doPut(blockId, IteratorValues(makeIterator), level, downgradeToReadLock = true) match {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I think we can simplify this by moving the acquiring lock logic outside of `doPut`. If we fail to acquire a write lock for the new block we used to return `None` in `doPut` anyway, so it would make sense to instead call `doPut` only if we're able to acquire the write lock. Then you don't need the complicated return type in `doPut`.

(If you don't want to change all the *other* places where `doPut` is called, we can just make an overloaded `doPut` method that takes in a `blockInfo` and assumes the write lock is already held.)

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Then this method will look something like:
```
val newInfo = new BlockInfo(level, tellMaster = true)
if (blockInfoManager.lockNewBlockForWriting(blockId, newInfo)) {
  doPut(blockId, IteratorValues(makeIterator), level)
  Left(getLocal(blockId).getOrElse { release lock and throw exception })
} else {
  logWarning(s"Block $blockId already exists on this machine; not re-adding it")
  Right(makeIterator)
}
```

**Comment by andrewor14** [comment]
@JoshRosen This looks good, but still somewhat complex. In particular, there are a few unintuitive signatures in `BlockManager` and related classes that I think can be improved. I made a suggestion for `doPut`'s return type, but I'm still thinking about how to avoid the `Either` in `getOrElseUpdate`'s return type. I think this task is inherently complex so eventually we'll run out of things to simplify without a significant refactor.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 12]
```diff
@@ -315,6 +315,34 @@ abstract class RDD[T: ClassTag](
...
+      computeOrReadCheckpoint(partition, context)
+    }) match {
+      case Left(blockResult) =>
+        if (readCachedBlock) {
+          val existingMetrics = context.taskMetrics().registerInputMetrics(blockResult.readMethod)
+          existingMetrics.incBytesReadInternal(blockResult.bytes)
+          new InterruptibleIterator[T](context, blockResult.data.asInstanceOf[Iterator[T]]) {
+            override def next(): T = {
+              existingMetrics.incRecordsReadInternal(1)
+              delegate.next()
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: crazy idea: what if we push all of this metrics code into `getOrElseUpdate`? Then we don't need the complicated left or right signature anymore and we can contain all the complexity within `BlockManager` itself. We might need to pass in `TaskContext` but I think it's pretty doable to make `getOrElseUpdate` return an `Iterator[T]`.

File: core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala (hunk 7 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala]: Andrew and I discussed this offline. The while loop here is actually necessary in the current code due to the way that lockForReading behaves when the block is removed by the writer, but I can make some changes in `lockForReading` itself which will allow me to simplify this. Stay tuned for an update.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 12 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: I considered this but had some reservations due to the fact that we might not want to consider every read of a BlockManager iterator to be a read of an input record (since we also read from iterators in the implementation of getSingle() and in a couple of other places, and those methods are used for things like TorrentBroadcast blocks which I don't think we want to treat as input records in for the purposes of these metrics).

On the other hand, this `getOrElseUpdate` method really only makes sense in the context of caching code anyways, so maybe it's fine to couple its semantics a little more tightly to its only current use.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 11 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Some of the inherent complexity here comes from the fact that `doPut()` can fail to cache a new block at `MEMORY_ONLY` because the block is too large to fully unroll and cannot be dropped to disk. In that case, `doPut()` needs to return back an iterator which chains the partially-unrolled values to the rest of the original iterator; that's why we have this weird return value in `doPut()` .

In the `else` branch in your code example, we need to handle the case where the block already exists by reading it from the block store, not by calling `makeIterator()` to compute it.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 5 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Done.

File: core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala (hunk 7 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala]: Updated; PTAL.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 8 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Good idea; updated.

File: core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala (hunk 7 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala]: well, after some refactoring you were able to get rid of the while loop after all :)

**Comment by andrewor14** [comment]
LGTM. There are still a few clunky interfaces but there's not much we can do about them without doing a bigger refactor, which is outside the scope.

**Comment by JoshRosen** [comment]
Yeah, I wish there was a straightforward way to simplify this code, but I think that a lot of the remaining complexity is pretty inherent in the current requirements / behaviors. I'm going to merge this in a little bit.

**Comment by andrewor14** [comment]
Merged into master.
